configuration/src/main/java/io/airlift/configuration/validation/FileExistsValidator.java

```diff
@@ -0,0 +1,57 @@
+package io.airlift.configuration.validation;
+
+import javax.validation.ConstraintValidator;
+import javax.validation.ConstraintValidatorContext;
+
+import java.io.File;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+import static java.lang.String.format;
+
+public class FileExistsValidator
+        implements ConstraintValidator<FileExists, Object>
+{
+    @Override
+    public void initialize(FileExists ignored)
+    {
+        // Annotation has no properties
+    }
+
+    @Override
+    public boolean isValid(Object path, ConstraintValidatorContext context)
+    {
+        if (path == null) {
+            // @NotNull's responsibility
+            return true;
+        }
+
+        boolean fileExists = exists(path);
+
+        if (!fileExists) {
+            context.disableDefaultConstraintViolation();
+            context.buildConstraintViolationWithTemplate(format("file '%s' does not exist", path.toString()))
+                    .addConstraintViolation();
+        }
+
+        return fileExists;
+    }
+
+    private static boolean exists(Object path)
+    {
+        if (path instanceof String) {
+            return Files.exists(Paths.get((String) path));
+        }
+
+        if (path instanceof Path) {
+            return Files.exists(((Path) path));
+        }
+
+        if (path instanceof File) {
+            return ((File) path).exists();
+        }
+
+        throw new IllegalStateException("Unsupported type: " + path.getClass().getName());
```

Comment:
This should be `IllegalArgumentException` since the argument is invalid. `IllegalStateException` means the class is in an incompatible state for the current method invocation.

Reply:
How about
```java
throw new IllegalArgumentException("Unsupported type for @FileExists: " + path.getClass().getName());